frontend/src/abs/frontend/pardef/ParFnAppParam.jrag

```diff
@@ -0,0 +1,173 @@
+    }
+
+    /**
+    * Adds the closure params for this anonymous function to the given FunctionDecl.
+    * If the params for this anonymous function's PureExp have already been added, this method will do nothing.
+    * After adding the closure params to the function, this AnonymousFunctionDecl will be added to the passed list.
+    *
+    * <p>If the function contains any recursive calls, the closure params will be added.</p>
+    *
+    * @param added a list of anonymous functions that already added their closure params
+    * @param func a function to add the closure params to
+    */
+    public void AnonymousFunctionDecl.addClosureParams(java.util.List<AnonymousFunctionDecl> added, FunctionDecl func) {
+        for (AnonymousFunctionDecl anon : added) {
+            if (anon.getPureExpNoTransform().getExpId().equals(this.getPureExpNoTransform().getExpId())) {
+                // Closure params for this pure expression were already added
+                return;
+            }
+        }
+
+        java.util.List<FnApp> recursiveCalls = new LinkedList<>();
+        for (FnApp call : func.findChildren(FnApp.class)) {
+            if (call.getName().equals(func.getName())) {
+                recursiveCalls.add(call);
+            }
+        }
+
+        List<ParamDecl> params = func.getParams();
+        for (ParamDecl freeVar : getFreeVars().values()) {
+            params.add(freeVar.treeCopyNoTransform());
+            for (FnApp recursiveCall : recursiveCalls) {
+                recursiveCall.addParam(new VarUse(freeVar.getName()));
+            }
+        }
+
+        added.add(this);
+    }
+
+    /**
+    * Adds the closure params for this anonymous function to the given ParFnApp.
+    * If the params for this anonymous function's PureExp have already been added, this method will do nothing.
+    * After adding the closure params to the function call, this AnonymousFunctionDecl will be added to the passed list.
+    * @param added a list of anonymous functions that already added their closure params
+    * @param func a function call to add the closure params to
+    */
+    public void AnonymousFunctionDecl.addClosureParams(java.util.List<AnonymousFunctionDecl> added, ParFnApp fnApp) {
+        for (AnonymousFunctionDecl anon : added) {
+            if (anon.getPureExpNoTransform().getExpId().equals(this.getPureExpNoTransform().getExpId())) {
+                // Closure params for this pure expression were already added
+                return;
+            }
+        }
+
+        for (String freeVarName : getFreeVars().keySet()) {
+            fnApp.addParam(new VarUse(freeVarName));
+        }
+
+        added.add(this);
+    }
+
+    // TODO this should probably be reset somewhere
+    /**
+    * Stores the IDs for all pure expressions. Keys are the pretty-printed pure expressions.
+    */
+    static Map<String, String> PureExp.pureExpIds = new HashMap<>(256);
+
```

Comment:
Could you run this by @rm2340 ? I could potentially see this static cache having an impact on the work he's doing in dynamic / multiple product generation.

Reply:
I realized that assigning IDs to pure expressions was the result of a premature optimization regarding anonymous functions. If an anonymous function is used multiple times, it should probably just be declared as a regular, named function.